src/mon/OSDMonitor.cc

```diff
@@ -5796,17 +5796,53 @@ bool OSDMonitor::prepare_command_impl(MMonCommand *m,
   } else if (prefix == "osd create") {
     int i = -1;
 
+    // optional id provided?
+    int64_t id = -1;
+    if (cmd_getval(g_ceph_context, cmdmap, "id", id)) {
+      if (id < 0) {
+	ss << "invalid osd id value '" << id << "'";
+	err = -EINVAL;
+	goto reply;
+      }
+      dout(10) << " osd create got id " << id << dendl;
+    } else {
+      id = -1;
+    }
+
     // optional uuid provided?
     uuid_d uuid;
     string uuidstr;
     if (cmd_getval(g_ceph_context, cmdmap, "uuid", uuidstr)) {
       if (!uuid.parse(uuidstr.c_str())) {
+	ss << "invalid uuid value '" << uuidstr << "'";
 	err = -EINVAL;
 	goto reply;
       }
       dout(10) << " osd create got uuid " << uuid << dendl;
       i = osdmap.identify_osd(uuid);
-      if (i >= 0) {
+      if (i < 0)
+	i = pending_inc.identify_osd(uuid);
+      if (id >= 0) {
+	// Both id and uuid provided. Check they do not conflict with existing.
+	if (i < 0) {
+	  if (osdmap.exists(id) || pending_inc.new_state.count(id)) {
+	    ss << "id " << id << " already in use and not match uuid " << uuid;
+	    err = -EINVAL;
+	    goto reply;
```

Comment:
Ok, I removed pending_inc.new_state.count(id) condition, so it will be caught later in:

```
if (pending_inc.new_state.count(i)) {
   // osd is about to exist
   wait_for_finished_proposal(new C_RetryMessage(this, m));
   return true;
}
```